avocado/utils/vmimage.py

```diff
@@ -0,0 +1,251 @@
+# This program is free software; you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation; either version 2 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+#
+# See LICENSE for more details.
+#
+# Copyright: Red Hat Inc. 2017
+# Author: Amador Pahim <apahim@redhat.com>
+
+"""
+Provides VM images acquired from official repositories
+"""
+
+import os
+import tempfile
+import uuid
+
+from . import asset
+from . import path as utils_path
+from . import process
+
+try:
+    import lzma
+    LZMA_CAPABLE = True
+except ImportError:
+    LZMA_CAPABLE = False
+
+
+class VMImageError(Exception):
+    """
+    Generic error class for VM Image exceptions
+    """
+
+
+class Image(object):
+
+    def __init__(self, name, major, minor, build, arch, checksum, algorithm,
+                 cache_dir, url):
+        """
+        Vm Image representation.
+
+        :param name: Name of the Operating System
+        :param major: Major part of the version
+        :param minor: Minor part of the version
+        :param build: Build identifier
+        :param arch: Architecture
+        :param checksum: Hash of the image, used to validate the download
+        :param algorithm: Algorithm of the provided checksum
+        :param url: URL with the filename of the image
+        """
+        self.name = name
+        self.major = major
+        self.minor = minor
+        self.build = build
+        self.arch = arch
+        self.checksum = checksum
+        self.algorithm = algorithm
+        if cache_dir is None:
+            cache_dir = tempfile.gettempdir()
+        self.cache_dir = cache_dir
+        self.url = url
+        self.username = None
+        self.password = None
+        self.path = None
+        self.get()
+
+    def get(self):
```

Comment:
Following the idea of making `Image` behave more like an `ImageProvider`, the `get()` method could look into the other attributes (major, minor, arch) and know how to get the image for the given provider.  The base implementation, as you exemplified here would use the asset fetcher, but in theory, it could be re-implemented differently by a given provider based on the info on a spec.

It could then, return an object similar to ImageSpec (or itself) with the missing attributes set:
 * user did not specify a checksum, but the provider knows about it
 * user did not bother choosing a "spin" or "flavor", but the provider picked a specific
 * the path to the actual downloaded image file ready to be used